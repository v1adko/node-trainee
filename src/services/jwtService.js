import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
