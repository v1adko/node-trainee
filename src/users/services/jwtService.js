const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
