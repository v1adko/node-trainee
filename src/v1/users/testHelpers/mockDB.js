import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
