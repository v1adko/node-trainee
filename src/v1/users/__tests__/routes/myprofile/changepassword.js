import simulate from '../../../../../tests/requestHelper';
import jwtService from '../../../../../services/jwtService';
import mockDB from '../../../testHelpers/mockDB';

const filename = __filename.slice(__dirname.length + 1, -3);

const USERNAME = `testUsername${filename}`;
const PASSWORD = `testPassword${filename}`;
const NEW_PASSWORD = `newTestPassword100${filename}`;
const WRONG_PASSWORD = `wrongTestPassword100${filename}`;
const WROUNG_USER_TOKEN = `wrongUserToken${filename}`;
let user;
let userId;
let userToken;

async function clean() {
  await mockDB.cleanDB();
  userId = null;
  userToken = null;
}

async function createUser() {
  user = await mockDB.createUser(USERNAME, PASSWORD);
  userId = user._id.toString();
  userToken = jwtService.generateJwt(user);
}

beforeAll(async () => mockDB.createDefaultUsers());

describe('Test the "/v1/myprofile/changepassword" path', () => {
  const route = '/v1/myprofile/changepassword';

  beforeEach(createUser);

  afterEach(clean);

  it('should change pass for user and return authentication status, valid token', async () => {
    const result = await simulate.put(
      route,
      200,
      { password: PASSWORD, newPassword: NEW_PASSWORD },
      userToken
    );

    const { auth, token } = result.body;
    const decodedToken = await jwtService.decoder(token);

    expect(auth).toBe(true);
    expect(decodedToken._id).toBe(userId);
  });

  it('should not change pass for user, because password is wrong', async () => {
    const result = await simulate.put(
      route,
      400,
      { password: WRONG_PASSWORD, newPassword: NEW_PASSWORD },
      userToken
    );
    const { message } = result.body;

    expect(message).toBe('Password is wrong');
  });

  it('should not change pass for user, because tokken is wrong', async () => {
    const result = await simulate.put(
      route,
      500,
      { password: PASSWORD, newPassword: NEW_PASSWORD },
      WROUNG_USER_TOKEN
    );
    const { message } = result.body;

    expect(message).toBe('Failed to authenticate token.');
  });
});
