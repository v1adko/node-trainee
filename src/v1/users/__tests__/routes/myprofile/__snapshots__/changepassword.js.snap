// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/myprofile/changepassword path should not change password because newPassword less than 6 symbols 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"newPassword\\" length must be at least 6 characters long",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/myprofile/changepassword path should not change password for user, because password is wrong 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): Password is wrong, please check input data.",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/myprofile/changepassword path should not change password for user, because tokken is wrong 1`] = `
Object {
  "code": 401,
  "message": "TokenValidationError(401): Invalid token, please repeat authentication.",
  "name": "TokenValidationError",
}
`;
